include/ddwaf.h

```diff
@@ -216,37 +215,33 @@ typedef void (*ddwaf_log_cb)(
  * @note If config is NULL, default values will be used, including the default
  *       free function (ddwaf_object_free).
  **/
-ddwaf_handle ddwaf_init(const ddwaf_object *rule,
+ddwaf_handle ddwaf_init(const ddwaf_object *ruleset,
     const ddwaf_config* config, ddwaf_ruleset_info *info);
 
 /**
- * ddwaf_destroy
+ * ddwaf_update
  *
- * Destroy a WAF instance.
+ * Update a ddwaf instance
  *
- * @param Handle to the WAF instance.
- */
-void ddwaf_destroy(ddwaf_handle handle);
-
-/**
- * ddwaf_update_rule_data
+ * @param rule ddwaf::object containing the patterns to be used by the WAF. (nonnull)
+ * @param info Optional ruleset parsing diagnostics. (nullable)
  *
- * Update existing rules with new rule data.
+ * @return Handle to the WAF instance.
  *
- * @param handle to the WAF instance.
- * @param data A ddwaf_object with the format [{id, type, [data]}].
- */
-DDWAF_RET_CODE ddwaf_update_rule_data(ddwaf_handle handle, ddwaf_object *data);
+ * @note If config is NULL, default values will be used, including the default
+ *       free function (ddwaf_object_free).
+ **/
+ddwaf_handle ddwaf_update(ddwaf_handle handle, const ddwaf_object *rule,
```

Comment:
This was updated while you were reviewing. 

> From this I can take that concurrent calls to ddwaf_update (even with different ddwaf_handles as they can share the same builder!) are forbidden and that ddwaf_update actually has side effects on the original handle.

Yes, indeed `ddwaf_update` isn't thread-safe regardless of the handle, I'll document it.

> Can the old handle still be used for non-update purposes? It seems to be case, because the ruleset is constructed anew.

Yes, it's still valid.

> Is synchronization needed to ensure that the memory state behind the new ddwaf_handle is fully visible on other threads when ddwaf_run is called? I guess the answer is no, so the bindings need at least an acquire/release when reading where the current ddwaf_handle is stored.

I think you'll only need to synchronize reads and writes around the shared `ddwaf_handle`.